pkg/metabase/grant.go

```diff
@@ -0,0 +1,397 @@
+package metabase
+
+import (
+	"context"
+	"database/sql"
+	"encoding/base64"
+	"errors"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/google/uuid"
+	"github.com/navikt/nada-backend/pkg/graph/models"
+	"google.golang.org/api/cloudresourcemanager/v1"
+	iam "google.golang.org/api/iam/v1"
+)
+
+func (m *Metabase) grantMetabaseAccess(ctx context.Context, dsID uuid.UUID, subject string) {
+	if subject == "group:all-users@nav.no" {
+		m.addAllUsersDataset(ctx, dsID)
+	} else if strings.HasPrefix(subject, "group:") {
+		m.addGroupAccess(ctx, dsID, subject)
+	} else {
+		m.addMetabaseGroupMember(ctx, dsID, subject)
+	}
+}
+
+func (m *Metabase) addAllUsersDataset(ctx context.Context, dsID uuid.UUID) {
+	log := m.log.WithField("datasetID", dsID)
+
+	mbMetadata, err := m.repo.GetMetabaseMetadata(ctx, dsID, true)
+	if errors.Is(err, sql.ErrNoRows) {
+		ds, err := m.repo.GetDataset(ctx, dsID)
+		if err != nil {
+			log.WithError(err).Error("getting dataset")
+			return
+		}
+
+		err = m.create(ctx, dsWrapper{
+			Dataset: ds,
+			Key:     m.sa,
+			Email:   m.saEmail,
+		})
+		if err != nil {
+			log.WithError(err).Error("creating metabase database")
+			return
+		}
+		return
+	} else if err != nil {
+		log.WithError(err).Error("get metabase metadata")
+		return
+	}
+
+	if mbMetadata.DeletedAt != nil {
+		log.Info("restoring db")
+		if err := m.restore(ctx, dsID, mbMetadata); err != nil {
+			log.WithError(err).Error("restoring db")
+		}
+		return
+	}
+	if mbMetadata.PermissionGroupID == 0 {
+		// all-users access exists
+		log.Info("all users database already exists in metabase")
+		return
+	} else {
+		if err := m.client.OpenAccessToDatabase(ctx, mbMetadata.DatabaseID); err != nil {
+			log.WithError(err).Error("open access to dataset")
+			return
+		}
+
+		if err := m.client.DeletePermissionGroup(ctx, mbMetadata.PermissionGroupID); err != nil {
+			log.WithError(err).Errorf("removing old permission group %v when opening database", mbMetadata.PermissionGroupID)
+		}
+
+		if err := m.repo.SetPermissionGroupMetabaseMetadata(ctx, mbMetadata.DatasetID, 0); err != nil {
+			log.WithError(err).Error("setting permission group to all users")
+			return
+		}
+	}
+}
+
+func (m *Metabase) addDatasetMapping(ctx context.Context, dsID uuid.UUID) {
+	log := m.log.WithField("datasetID", dsID)
+
+	mbMeta, err := m.repo.GetMetabaseMetadata(ctx, dsID, true)
+	if errors.Is(err, sql.ErrNoRows) {
+		ds, err := m.repo.GetDataset(ctx, dsID)
+		if err != nil {
+			log.WithError(err).Error("getting dataproduct")
+			return
+		}
+
+		if err := m.createRestricted(ctx, ds); err != nil {
+			log.WithError(err).Error("create restricted database")
+			return
+		}
+
+		if err := m.grantAccessesOnCreation(ctx, dsID); err != nil {
```

Comment:
This if clause and line 121-122 seems mergable if move them below line 126 and remove line 103, 125
But it is just a personal opinion